### PR TITLE
Add `clear_task_if_completed` function to `WorkerThreadPool`

### PR DESCRIPTION
**Pull request by BrianBHuynh**
Currently every task must be waited on (wait_for_task_completion) for their resources to be removed from memory (https://github.com/godotengine/godot/issues/84888). This is addressed in the docs however said solution may not always be intuitive or ideal as the function known as "wait_for_task_completion" is used even in cases where no waiting is necessary as it does the double duty of waiting, and also clearing memory.

While it makes sense that if you wait for a task, and it finishes, it should be cleared from memory, it is not always intuitive that you HAVE to wait for a task that you know is completed (via is_task_completed) to prevent a memory leak.

This can currently be circumvented by using an if statement (if is_task_completed: wait_for_task_completion) however said usage is ugly and also unintuitive.

This commit adds a function which checks if the task is completed, and then returns true and clears resources if so. Essentially doing the same job as wait_for_task_completion if already completed, HOWEVER not binding the main thread (or the thread it is called from) in the case the task isn't.

This allows for resource clearing without accidentally pausing the entire thread if not completed, in the case of a non essential functions such as saving (which can be allowed to operate in the background without pausing the entire game/main thread)

### CODE REVIEW

**Comment by BrianBHuynh** [comment]
Just merged the commit since most of the test seemed to have been working after the small fix :D

**Comment by BrianBHuynh** [comment]
Ah I need to update the documentation, 1 second

**Comment by RadiantUwU** [comment]
Why not go overboard, make a new function named `clear_task_when_complete` which will automatically handle cleaning it out of memory? Sure, you can't wait on the task anymore, and you cant check when it finishes, but neither this function as well.

File: core/object/worker_thread_pool.cpp [hunk 1]
```diff
@@ -361,6 +361,26 @@ bool WorkerThreadPool::is_task_completed(TaskID p_task_id) const {
 	return (*taskp)->completed;
 }
 
+bool WorkerThreadPool::clear_task_if_completed(TaskID p_task_id) {
+	task_mutex.lock();
+	Task **taskp = tasks.getptr(p_task_id);
+	if (!taskp) {
+		task_mutex.unlock();
```
Comment (RadiantUwU) [inline core/object/worker_thread_pool.cpp:368]: If the task has been already waited on, it will raise an error.

**Comment by BrianBHuynh** [comment]
I was thinking it'd be nice to have a way for the developer to manually clear a task (and acknowledge that it was finished via the bool) without having to call the wait function, but yeah I think it's a good idea to implement a auto clear task function like you suggested. 

Atleast for me I think having something like this would be nice in general though in the case the programmer needs to acknowledge the task being completed in a way which isn't as time sensitive as using a whole wait for.

I'll work on implementing a auto clear function as well though

File: core/object/worker_thread_pool.cpp (hunk 1 above)
Comment (BrianBHuynh) [reply core/object/worker_thread_pool.cpp:368]: Oh yeah I guess that lock isn't needed huh

Comment (BrianBHuynh) [reply core/object/worker_thread_pool.cpp]: Just removed the locks!

**Comment by RandomShaper** [comment]
I'm not quite sure this adds significant value. On the one hand, the current API is designed to mimic joining threads. Being that the case, code using the threadpool would be designed in a way that the need for something like this is avoided in the first place. On the other hand, while there may be scenarios where this may be useful, given how wary we should be of growing the engine, let alone its core, there should be a proposal, backed by a high amount of real-world demand. Other approaches would be discussed there. Also, maybe other ideas such as renaming `wait` to `join` could be scheduled for future releases, for instance.

**Comment by BrianBHuynh** [comment]
Another idea I just thought of was combining the is_completed function that was already there with the function I wrote, giving a bool (or possibly just clearing the memory on true) which would dictate if the task would be cleared from memory as well!

Wanted to write it here as well since I just thought of it haha

**Comment by RandomShaper** [comment]
All those ideas may have merit, but the point about needing real-world examples and discussing alternatives and roadmap in a proposal stands.

**Comment by BrianBHuynh** [comment]
Alright! I'll start working on writing a proposal for the ideas I have and see what people think of them then! Thankyou for reading over what I had so far (I know it's alot haha)

**Pull request closed without merging**